app/models/manageiq/providers/redhat/infra_manager/vm.rb

```diff
@@ -48,17 +48,18 @@ def parent_cluster
   alias ems_cluster parent_cluster
 
   def disconnect_storage(_s = nil)
-    unless active?
-      return
-    end
+    return unless active?
+
+    vm_storages = ([storage] + storages).compact.uniq
+    return if vm_storages.empty?
+
     vm_disks = collect_disks
 
-    if vm_disks.blank?
-      storage = nil
-    else
-      vm_storages = ([storage] + storages).compact.uniq
-      storage = vm_storages.select { |store| !vm_disks.include?(store.ems_ref) }
-    end
+    storage = if vm_disks.blank?
```

Comment:
I don't think this behave the same as the code that @pkliczewski wrote.
In @pkliczewski there will be an assignment to `storage` in any case.
In your suggestion, the `storage` might have a previous value that won't be erased by your suggested fix.